### PR TITLE
monitoring-example

### PR DESCRIPTION
**Pull request by ahmadiesa-abu**
adding monitoring  plugin with example

### CODE REVIEW

File: cloudify-web-monitoring-plugin/README.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
+cloudify-web-monitoring-plugin
+===================
+
+
+# Description
+
+A Cloudify Plugin that monitor a URL response time and based on metrics will trigger scale up or down
```
Comment (EarthmanT) [inline cloudify-web-monitoring-plugin/README.md:7]:
```suggestion
A Cloudify Plugin that monitors a URL response time based on metrics will trigger scale up or down.
```

File: cloudify-web-monitoring-plugin/README.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+Basic contifuration includes what to monitor (URL), thresholds, limits and how to perform the scaling. Please see node type description below for details
```
Comment (EarthmanT) [inline cloudify-web-monitoring-plugin/README.md:9]:
```suggestion
Basic configuration that includes what to monitor (URL), thresholds, limits and how to perform the scaling. Please see node type description below for details
```

File: cloudify-web-monitoring-plugin/README.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+- url - URL to monitor the response time for
+- low_threshold - url response miliseconds threshold below which a scale down will be triggered (if there is more than 1 instance)
+- high_threshold - url response miliseconds threshold above which a scale up will be triggered (if the scaleout_limit is not reached)
+- scaleup_cooldown - minutes to wait in between scaleups
+- scaledown_cooldown - minutes to wait in between scaledowns
+- interval - minutes URL monitoring interval. The workflow that performs URL check will be scheduled with this rate
+- scaleout_limit - maximum number of scale out workflows to execute in a sequence e.g. when scale out is called a counter is incremented and compared to this value. When scale down is called the counter is decremented
+- client - optional Cloudfiy REST configuration
+
+## Building the plugin
```
Comment (EarthmanT) [inline cloudify-web-monitoring-plugin/README.md:27]: Let's skip the local build - just use the wagon builder. Also instruct them how to acquire the container. I think the ones on Docker hub are way out of date.

File: cloudify-web-monitoring-plugin/blueprints/AKS/README.md [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+### Creating secrets
+
+Replace <value> with actual values, without the <>
+
+```shell
+cfy secrets create azure_subscription_id --secret-string <value>
+cfy secrets create azure_tenant_id --secret-string <value>
+cfy secrets create azure_client_id --secret-string <value>
+cfy secrets create azure_client_secret --secret-string <value>
+```
```
Comment (EarthmanT) [inline cloudify-web-monitoring-plugin/blueprints/AKS/README.md:25]: Can you point to documentation where the user can find these secrets?

File: cloudify-web-monitoring-plugin/blueprints/AKS/cluster.yaml [hunk 5]
```diff
@@ -0,0 +1,386 @@
+tosca_definitions_version: cloudify_dsl_1_3
```
Comment (EarthmanT) [inline cloudify-web-monitoring-plugin/blueprints/AKS/cluster.yaml:1]: No I don't think so! Can we use the existing Kubernetes examples that we already have to support? I don't want to support more Kubernetes examples. Support yet another? No freaking way!

File: cloudify-web-monitoring-plugin/blueprints/AKS/monitor.yaml [hunk 6]
```diff
@@ -0,0 +1,36 @@
+tosca_definitions_version: cloudify_dsl_1_3
```
Comment (EarthmanT) [inline cloudify-web-monitoring-plugin/blueprints/AKS/monitor.yaml:1]: This is fine.

File: cloudify-web-monitoring-plugin/blueprints/AKS/resources/template.yaml [hunk 7]
```diff
@@ -0,0 +1,32 @@
+apiVersion: v1
```
Comment (EarthmanT) [inline cloudify-web-monitoring-plugin/blueprints/AKS/resources/template.yaml:1]: This is fine.

File: cloudify-web-monitoring-plugin/blueprints/AKS/scripts/get_agent_pool.py [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    }
+    response = requests.post(url, data=body)
+    return response.json()
+
+def get_agent_pool():
+    azure_subscription_id = inputs['azure_config']['subscription_id']
+    resource_group_name = inputs['resource_group_name']
+    cluster_name = inputs['cluster_name']
+    agent_pool_name = inputs['agent_pool_name']
+    url = "https://management.azure.com/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ContainerService/managedClusters/{resourceName}/agentPools/{agentPoolName}?api-version=2021-03-01".format(
```
Comment (EarthmanT) [inline cloudify-web-monitoring-plugin/blueprints/AKS/scripts/get_agent_pool.py:38]: I'm sure there is another way to get this value, for example from the runtime properties of the existing kubernetes cluster. By the way, if not, we can add it as an output to the existing blueprint.

File: cloudify-web-monitoring-plugin/blueprints/AKS/scripts/get_agent_pool.py [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+            subscriptionId = azure_subscription_id,
+            resourceGroupName = resource_group_name,
+            resourceName = cluster_name,
+            agentPoolName = agent_pool_name
+        )
+    headers = {
+        'Authorization': 'Bearer ' + authorize_with_azure()['access_token'],
```
Comment (EarthmanT) [inline cloudify-web-monitoring-plugin/blueprints/AKS/scripts/get_agent_pool.py:45]: Existing blueprint already has token as an output.

File: cloudify-web-monitoring-plugin/blueprints/AKS/scripts/scale_nodes.py [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+        )
+    headers = {
+        'Authorization': 'Bearer ' + authorize_with_azure()['access_token'],
+        'Content-Type': 'application/json'
+        }
+    response = requests.get(url, headers=headers)
+    ctx.instance.runtime_properties['agent_pool'] = response.json()
+    ctx.instance.runtime_properties['node_count'] = \
+        response.json()['properties']['count']
+    return response.json()
```
Comment (EarthmanT) [inline cloudify-web-monitoring-plugin/blueprints/AKS/scripts/scale_nodes.py:54]: Why duplicating code from another script?

File: cloudify-web-monitoring-plugin/blueprints/AKS/scripts/store_kube_token.py [hunk 11]
```diff
@@ -0,0 +1,11 @@
...
+
+from cloudify import ctx
+from cloudify.state import ctx_parameters as inputs
+from cloudify.manager import get_rest_client
+
+
+client = get_rest_client()
+token = base64.b64decode(inputs['kube_token']).decode('utf-8')
+client.secrets.create('kubernetes_token',  token, update_if_exists=True)
+ctx.instance.runtime_properties['token'] = token
```
Comment (EarthmanT) [inline cloudify-web-monitoring-plugin/blueprints/AKS/scripts/store_kube_token.py:11]: We already have this in the other examples....

**Review by EarthmanT — APPROVED** [review]

**Review by EarthmanT — CHANGES_REQUESTED** [review]
See requested changes.

File: cloudify-web-monitoring-plugin/README.md (hunk 3 above)
Comment (ahmadiesa-abu) [reply cloudify-web-monitoring-plugin/README.md:27]: added a reference to wagon builders

File: cloudify-web-monitoring-plugin/blueprints/AKS/README.md (hunk 4 above)
Comment (ahmadiesa-abu) [reply cloudify-web-monitoring-plugin/blueprints/AKS/README.md:27]: added a link to azure plugin documenation

File: cloudify-web-monitoring-plugin/blueprints/AKS/scripts/scale_nodes.py (hunk 10 above)
Comment (ahmadiesa-abu) [reply cloudify-web-monitoring-plugin/blueprints/AKS/scripts/scale_nodes.py:54]: removed the other script no need for it

**Comment by ahmadiesa-abu** [comment]
@EarthmanT Changes were applied

File: cloudify-web-monitoring-plugin/blueprints/AKS/cluster.yaml (hunk 5 above)
Comment (ahmadiesa-abu) [reply cloudify-web-monitoring-plugin/blueprints/AKS/cluster.yaml:1]: Changed to use AKS example as component

File: cloudify-web-monitoring-plugin/README.md [hunk 12]
```diff
@@ -26,16 +26,14 @@ Basic contifuration includes what to monitor (URL), thresholds, limits and how t
...
-wagon create -r dev-requirements.txt --build-tag "centos-Core" -v -f .
-```
+you can check this repository for more info [Cloudify wagon builders](https://github.com/cloudify-cosmo/cloudify-wagon-build-containers)
 
-or you can use Cloudify wagon builder Docker images
+after building the docker container , for example centos-7-py3 
 
 ```
-docker run -v {path_to_plugin}/:/packaging cloudifyplatform/cloudify-centos-7-py3-wagon-builder
+docker run -v {path_to_plugin}/:/packaging cloudify-centos-7-py3-wagon-builder
```
Comment (EarthmanT) [inline cloudify-web-monitoring-plugin/README.md:36]: `docker run -v ./web-monitoring-solution/plugin/:/packaging cloudify-centos-7-py3-wagon-builder `

File: cloudify-web-monitoring-plugin/README.md [hunk 13]
```diff
@@ -4,9 +4,9 @@ cloudify-web-monitoring-plugin
```
Comment (EarthmanT) [inline cloudify-web-monitoring-plugin/README.md:4]: Let's change the name of the folder from `cloudify-web-monitoring-plugin` to `web-monitoring-solution`

File: cloudify-web-monitoring-plugin/plugin.yaml [hunk 14]
```diff
@@ -0,0 +1,69 @@
+plugins:
```
Comment (EarthmanT) [inline cloudify-web-monitoring-plugin/plugin.yaml:1]: Let's put all plugin files in a subfolder named "plugin".

**Review by EarthmanT — COMMENTED** [review]
Made some comments. What about:

- aws
- openstack
- gcp

File: cloudify-web-monitoring-plugin/plugin.yaml (hunk 14 above)
Comment (ahmadiesa-abu) [reply cloudify-web-monitoring-plugin/plugin.yaml]: done

File: cloudify-web-monitoring-plugin/README.md (hunk 13 above)
Comment (ahmadiesa-abu) [reply cloudify-web-monitoring-plugin/README.md]: done

**Pull request merged**